Sources/SymbolKit/SymbolGraph/SemanticVersion/Prerelease.swift

```diff
@@ -0,0 +1,158 @@
+    /// - Parameter identifiers: The given pre-release identifier strings to create a semantic version pre-release from.
+    /// - Throws: A `SymbolGraph.SemanticVersionError` instance if any element of `identifiers` is not a valid pre-release identifier string.
+    internal init<C: Collection, S: StringProtocol>(_ identifiers: C) throws where C.Element == S, S.SubSequence == Substring {
+        self.identifiers = try identifiers.map {
+            try Identifier($0)
+        }
+    }
+}
+
+extension SymbolGraph.SemanticVersion.Prerelease.Identifier {
+    /// Creates a semantic version pre-release identifier from the given pre-release identifier string.
+    /// - Parameter identifierString: The given pre-release identifier string to create a semantic version pre-release identifier from.
+    /// - Throws: A `SymbolGraph.SemanticVersionError` instance if `identifierString` is not a valid pre-release identifier string.
+    internal init<S: StringProtocol>(_ identifierString: S) throws where S.SubSequence == Substring {
+        guard !identifierString.isEmpty else {
+            throw SymbolGraph.SemanticVersionError.emptyIdentifier(position: .prerelease)
+        }
+        guard identifierString.allSatisfy(\.isAllowedInSemanticVersionIdentifier) else {
+            throw SymbolGraph.SemanticVersionError.invalidCharacterInIdentifier(String(identifierString), position: .prerelease)
+        }
+        if identifierString.allSatisfy(\.isNumber) {
+            // diagnose the identifier as a numeric identifier, if all characters are ASCII digits
+            guard identifierString.first != "0" || identifierString == "0" else {
+                throw SymbolGraph.SemanticVersionError.invalidNumericIdentifier(
+                    String(identifierString),
+                    position: .prerelease,
+                    errorKind: .leadingZeros
+                )
+            }
+            guard let numericIdentifier = UInt(identifierString) else {
+                if identifierString.isEmpty {
+                    throw SymbolGraph.SemanticVersionError.emptyIdentifier(position: .prerelease)
+                } else {
+                    throw SymbolGraph.SemanticVersionError.invalidNumericIdentifier(
+                        String(identifierString),
+                        position: .prerelease,
+                        errorKind: .oversizedValue
+                    )
+                }
+            }
+            self = .numeric(numericIdentifier)
+        } else {
+            self = .alphanumeric(String(identifierString))
+        }
+    }
+}
+
+// MARK: - Comparable Conformances
+
+// Compiler synthesised `Equatable`-conformance is correct here.
+extension SymbolGraph.SemanticVersion.Prerelease: Comparable {
+    internal static func <(lhs: Self, rhs: Self) -> Bool {
+        guard !lhs.identifiers.isEmpty else { return false } // non-pre-release lhs >= potentially pre-release rhs
```

Comment:
True, but then we'd be trading an empty check for an optional check. We always have to check whether there is pre-release information regardless. Additionally, it will necessitate re-parsing (and maybe validating) the pre-release string every time on comparison.